### PR TITLE
Add temporal sampling support to `NeighborLoader`

### PR DESCRIPTION
**Pull request by RexYing**
not sure how to use the parameterize thing yet.

I was trying 
@pytest.mark.parametrize('directed', [True, False])
@pytest.mark.parametrize('temporal', [True, False])
but didn't work. Will look into it further.

### CODE REVIEW

**Comment by rusty1s** [comment]
> I was trying
> @pytest.mark.parametrize('directed', [True, False])
> @pytest.mark.parametrize('temporal', [True, False])
> but didn't work. Will look into it further.

This should work, but keep in mind to add both `directed` and `temporal` to the function signature.

File: torch_geometric/loader/neighbor_loader.py [hunk 1]
```diff
@@ -207,6 +225,7 @@ def __init__(
         directed: bool = True,
         transform: Callable = None,
         neighbor_sampler: Optional[NeighborSampler] = None,
+        node_time: Optional[Dict[NodeType, Tensor]] = None,
```
Comment (Padarn) [inline torch_geometric/loader/neighbor_loader.py:228]: Can we add a docstring about the new functionality?

Comment (RexYing) [reply torch_geometric/loader/neighbor_loader.py:238]: done.

File: torch_geometric/data/lightning_datamodule.py [hunk 2]
```diff
@@ -201,7 +201,7 @@ class LightningNodeData(LightningDataModule):
             indices of test nodes. If not given, will try to automatically
             infer them from the :obj:`data` object. (default: :obj:`None`)
         loader (str): The scalability technique to use (:obj:`"full"`,
-            :obj:`"neighbor"`). (default: :obj:`"neighbor"`)
+            :obj:`"neighbor"`, :obj:`"temporal"`). (default: :obj:`"neighbor"`)
```
Comment (rusty1s) [inline torch_geometric/data/lightning_datamodule.py:204]: As discussed: Let's remove this. `time_attr` can be set within `kwargs`.

File: torch_geometric/loader/neighbor_loader.py [hunk 3]
```diff
@@ -25,19 +25,24 @@ def __init__(
...
         self.replace = replace
         self.directed = directed
+        self.node_time = None
 
         if isinstance(data, Data):
             # Convert the graph data into a suitable format for sampling.
             out = to_csc(data, device='cpu', share_memory=share_memory)
             self.colptr, self.row, self.perm = out
             assert isinstance(num_neighbors, (list, tuple))
+            # TODO: time_attr for homogeneous graphs
```
Comment (rusty1s) [inline torch_geometric/loader/neighbor_loader.py:41]: Is it possible to fail here, by raising a `ValueError`?

File: torch_geometric/loader/neighbor_loader.py [hunk 4]
```diff
@@ -78,18 +83,34 @@ def __call__(self, index: Union[List[int], Tensor]):
...
+                    self.row_dict,
+                    {self.input_type: index},
+                    self.num_neighbors,
+                    self.num_hops,
+                    self.replace,
+                    self.directed,
+                )
+            else:
+                sample_fn = \
+                    torch.ops.torch_sparse.hetero_neighbor_temporal_sample
```
Comment (rusty1s) [inline torch_geometric/loader/neighbor_loader.py:101]:
```suggestion
                    torch.ops.torch_sparse.hetero_temporal_neighbor_sample
```

File: torch_geometric/loader/neighbor_loader.py [hunk 5]
```diff
@@ -25,19 +25,24 @@ def __init__(
...
         if isinstance(data, Data):
             # Convert the graph data into a suitable format for sampling.
             out = to_csc(data, device='cpu', share_memory=share_memory)
             self.colptr, self.row, self.perm = out
             assert isinstance(num_neighbors, (list, tuple))
+            # TODO: time_attr for homogeneous graphs
 
         elif isinstance(data, HeteroData):
+            if time_attr is not None:
+                self.node_time = data[time_attr]
```
Comment (rusty1s) [inline torch_geometric/loader/neighbor_loader.py:45]: This should be a dict, right? I think we need to do `self.node_time_dict = data.collect(time_attr)` instead.

File: torch_geometric/loader/neighbor_loader.py [hunk 6]
```diff
@@ -212,6 +233,16 @@ class NeighborLoader(torch.utils.data.DataLoader):
         transform (Callable, optional): A function/transform that takes in
             a sampled mini-batch and returns a transformed version.
             (default: :obj:`None`)
+        time_attr (str, optional): An optional argument
+            to specify the node attribute that denotes the timestamps for
+            nodes in graph.
+            If set, data[time_attr] has to be a dictionary of type
```
Comment (rusty1s) [inline torch_geometric/loader/neighbor_loader.py:239]: This is slightly off. Each node type sets its own `time_attr`, e.g., `data[node_type].time = ...`.

**Review by rusty1s — CHANGES_REQUESTED** [review]
This looks pretty good! Left some last comments.

**Pull request merged**